File: go/0159_longest_substring_with_at_most_two_distinct_characters.go
func lengthOfLongestSubstringTwoDistinct(s string) int {
	slen := len(s)
	if slen <= 2 {
		return slen
	}

	left, right, max := 0, 0, 0
	m := make(map[byte]int)
	for {
		if len(m) > 2 {
			for {
				count := m[s[left]]
				if count == 1 {
					delete(m, s[left])
					if len(m) <= 2 {
						left++
						break
					}
				} else {
					m[s[left]] = count - 1
				}
				left++
			}
		} else {
			count, ok := m[s[right]]
			if !ok {
				m[s[right]] = 1
			} else {
				m[s[right]] = count + 1
			}

			if len(m) <= 2 && right-left+1 > max {
				max = right - left + 1
			}
			right++
			if right == len(s) {
				break
			}
		}
	}
	return max
}

File: python/0159_longest_substring_with_at_most_two_distinct_characters.py
from collections import defaultdict

class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s):
        """
        :type s: str
        :rtype: int
        """

        mp = defaultdict(lambda: 0)
        res = 0

        left, right = 0, 0
        while len(mp) <= 2 and right <= len(s) - 1:
            mp[s[right]] += 1
            if len(mp) <= 2 and right - left + 1 > res:
                res = right - left + 1
            right += 1

        while right <= len(s) - 1:
            while len(mp) >= 3:
                mp[s[left]] -= 1
                if mp[s[left]] == 0:
                    mp.pop(s[left], None)
                left += 1

            while len(mp) <= 2 and right <= len(s) - 1:
                mp[s[right]] += 1
                if len(mp) <= 2 and right - left + 1 > res:
                    res = right - left + 1
                right += 1
        return res
